Lib/threading.py

```diff
@@ -980,15 +987,24 @@ def start(self):
         if self._started.is_set():
             raise RuntimeError("threads can only be started once")
 
+        self._join_lock = _allocate_lock()
+
         with _active_limbo_lock:
             _limbo[self] = self
         try:
-            _start_new_thread(self._bootstrap, ())
+            # Start joinable thread
+            _start_new_thread(self._bootstrap, (), {}, True)
         except Exception:
             with _active_limbo_lock:
                 del _limbo[self]
             raise
-        self._started.wait()
+        self._started.wait()  # Will set ident and native_id
+
+        # We need to make sure the OS thread is either explicitly joined or
+        # detached at some point, otherwise system resources can be leaked.
+        def _finalizer(wr, _detach_thread=_detach_thread, ident=self._ident):
+            _detach_thread(ident)
+        self._non_joined_finalizer = _weakref.ref(self, _finalizer)
```

Comment:
Er... yes, silly me. This would need to be stored in a global finalization registry in order to have any effect, which is basically what `weakref.finalize` does (but we don't want to import the whole `weakref` module here).

I think I'll just switch to a plain `__del__` method. I'm not sure why I was trying to avoid it.

Reply:
> I think I'll just switch to a plain `__del__` method. I'm not sure why I was trying to avoid it.

PTSD I'm guessing. 😄

Reply:
The finalization is now done on the thread handle object.